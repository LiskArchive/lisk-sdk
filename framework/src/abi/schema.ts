/*
 * Copyright Â© 2022 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import {
	blockHeaderSchema,
	blockSchema,
	blockAssetSchema,
	transactionSchema,
} from '@liskhq/lisk-chain';

export { blockHeaderSchema, blockSchema };

export const eventSchema = {
	$id: '/block/event',
	type: 'object',
	required: ['module', 'typeID', 'data', 'topics', 'index'],
	properties: {
		module: {
			dataType: 'string',
			fieldNumber: 1,
		},
		typeID: {
			dataType: 'bytes',
			fieldNumber: 2,
		},
		data: {
			dataType: 'bytes',
			fieldNumber: 3,
		},
		topics: {
			type: 'array',
			fieldNumber: 4,
			items: {
				maxItems: 4,
				dataType: 'bytes',
			},
		},
		index: {
			dataType: 'uint32',
			fieldNumber: 5,
		},
	},
};

export const validatorSchema = {
	$id: '/abi/validator',
	type: 'object',
	required: ['address', 'bftWeight', 'generatorKey', 'blsKey'],
	properties: {
		address: {
			fieldNumber: 1,
			dataType: 'bytes',
		},
		bftWeight: {
			fieldNumber: 2,
			dataType: 'uint64',
		},
		generatorKey: {
			fieldNumber: 3,
			dataType: 'bytes',
		},
		blsKey: {
			fieldNumber: 4,
			dataType: 'bytes',
		},
	},
};

export const consensusSchema = {
	$id: '/abi/consensus',
	type: 'object',
	required: ['currentValidators', 'implyMaxPrevote', 'maxHeightCertified'],
	properties: {
		currentValidators: {
			fieldNumber: 1,
			type: 'array',
			items: {
				...validatorSchema,
			},
		},
		implyMaxPrevote: {
			fieldNumber: 2,
			dataType: 'boolean',
		},
		maxHeightCertified: {
			fieldNumber: 3,
			dataType: 'uint32',
		},
		certificateThreshold: {
			fieldNumber: 4,
			dataType: 'uint64',
		},
	},
};

export const readyRequestSchema = {
	$id: '/abi/readyRequest',
	type: 'object',
	required: ['chainID', 'lastBlockHeight'],
	properties: {
		chainID: {
			fieldNumber: 1,
			dataType: 'bytes',
		},
		lastBlockHeight: {
			fieldNumber: 2,
			dataType: 'uint32',
		},
	},
};

export const readyResponseSchema = {
	$id: '/abi/readyResponse',
	type: 'object',
	required: [],
	properties: {},
};

export const initStateMachineRequestSchema = {
	$id: '/abi/initStateMachineRequest',
	type: 'object',
	required: ['header'],
	properties: {
		header: {
			fieldNumber: 1,
			...blockHeaderSchema,
		},
	},
};

export const initStateMachineResponseSchema = {
	$id: '/abi/initStateMachineResponse',
	type: 'object',
	required: ['contextID'],
	properties: {
		contextID: {
			fieldNumber: 1,
			dataType: 'bytes',
		},
	},
};

export const initGenesisStateRequestSchema = {
	$id: '/abi/initGenesisStateRequest',
	type: 'object',
	required: ['contextID'],
	properties: {
		contextID: {
			fieldNumber: 1,
			dataType: 'bytes',
		},
	},
};

export const initGenesisStateResponseSchema = {
	$id: '/abi/initGenesisStateResponse',
	type: 'object',
	required: ['events', 'preCommitThreshold', 'certificateThreshold', 'nextValidators'],
	properties: {
		events: {
			fieldNumber: 1,
			type: 'array',
			items: {
				...eventSchema,
			},
		},
		preCommitThreshold: {
			fieldNumber: 2,
			dataType: 'uint64',
		},
		certificateThreshold: {
			fieldNumber: 3,
			dataType: 'uint64',
		},
		nextValidators: {
			fieldNumber: 4,
			type: 'array',
			items: {
				...validatorSchema,
			},
		},
	},
};

export const insertAssetsRequestSchema = {
	$id: '/abi/insertAssetsRequest',
	type: 'object',
	required: ['contextID', 'finalizedHeight'],
	properties: {
		contextID: {
			fieldNumber: 1,
			dataType: 'bytes',
		},
		finalizedHeight: {
			fieldNumber: 2,
			dataType: 'uint32',
		},
	},
};

export const insertAssetsResponseSchema = {
	$id: '/abi/insertAssetsResponse',
	type: 'object',
	required: ['assets'],
	properties: {
		assets: {
			fieldNumber: 1,
			type: 'array',
			items: {
				...blockAssetSchema,
			},
		},
	},
};

export const verifyAssetsRequestSchema = {
	$id: '/abi/verifyAssetsRequest',
	type: 'object',
	required: ['contextID', 'assets'],
	properties: {
		contextID: {
			fieldNumber: 1,
			dataType: 'bytes',
		},
		assets: {
			fieldNumber: 2,
			type: 'array',
			items: {
				...blockAssetSchema,
			},
		},
	},
};

export const verifyAssetsResponseSchema = {
	$id: '/abi/verifyAssetsResponse',
	type: 'object',
	properties: {},
};

export const beforeTransactionsExecuteRequestSchema = {
	$id: '/abi/beforeTransactionsExecuteRequest',
	type: 'object',
	required: ['contextID', 'assets', 'consensus'],
	properties: {
		contextID: {
			fieldNumber: 1,
			dataType: 'bytes',
		},
		assets: {
			fieldNumber: 2,
			type: 'array',
			items: {
				...blockAssetSchema,
			},
		},
		consensus: {
			fieldNumber: 3,
			...consensusSchema,
		},
	},
};

export const beforeTransactionsExecuteResponseSchema = {
	$id: '/abi/beforeTransactionsExecuteResponse',
	type: 'object',
	required: ['events'],
	properties: {
		events: {
			fieldNumber: 1,
			type: 'array',
			items: {
				...eventSchema,
			},
		},
	},
};

export const afterTransactionsExecuteRequestSchema = {
	$id: '/abi/afterTransactionsExecuteRequest',
	type: 'object',
	required: ['contextID', 'assets', 'consensus', 'transactions'],
	properties: {
		contextID: {
			fieldNumber: 1,
			dataType: 'bytes',
		},
		assets: {
			fieldNumber: 2,
			type: 'array',
			items: {
				...blockAssetSchema,
			},
		},
		consensus: {
			fieldNumber: 3,
			...consensusSchema,
		},
		transactions: {
			fieldNumber: 4,
			type: 'array',
			items: {
				...transactionSchema,
			},
		},
	},
};

export const afterTransactionsExecuteResponseSchema = {
	$id: '/abi/afterTransactionsExecuteResponse',
	type: 'object',
	required: ['events', 'preCommitThreshold', 'certificateThreshold', 'nextValidators'],
	properties: {
		events: {
			fieldNumber: 1,
			type: 'array',
			items: {
				...eventSchema,
			},
		},
		preCommitThreshold: {
			fieldNumber: 2,
			dataType: 'uint64',
		},
		certificateThreshold: {
			fieldNumber: 3,
			dataType: 'uint64',
		},
		nextValidators: {
			fieldNumber: 4,
			type: 'array',
			items: {
				...validatorSchema,
			},
		},
	},
};

export const verifyTransactionRequestSchema = {
	$id: '/abi/verifyTransactionRequest',
	type: 'object',
	required: ['contextID', 'transaction'],
	properties: {
		contextID: {
			fieldNumber: 1,
			dataType: 'bytes',
		},
		transaction: {
			fieldNumber: 2,
			...transactionSchema,
		},
	},
};

export const verifyTransactionResponseSchema = {
	$id: '/abi/verifyTransactionResponse',
	type: 'object',
	required: ['result'],
	properties: {
		result: {
			fieldNumber: 1,
			dataType: 'sint32',
		},
	},
};

export const executeTransactionRequestSchema = {
	$id: '/abi/executeTransactionRequest',
	type: 'object',
	required: ['contextID', 'transaction', 'assets', 'dryRun', 'header'],
	properties: {
		contextID: {
			fieldNumber: 1,
			dataType: 'bytes',
		},
		transaction: {
			fieldNumber: 2,
			...transactionSchema,
		},
		assets: {
			fieldNumber: 3,
			type: 'array',
			items: {
				...blockAssetSchema,
			},
		},
		dryRun: {
			fieldNumber: 4,
			dataType: 'boolean',
		},
		header: {
			fieldNumber: 5,
			...blockHeaderSchema,
		},
		consensus: {
			fieldNumber: 6,
			...consensusSchema,
		},
	},
};

export const executeTransactionResponseSchema = {
	$id: '/abi/executeTransactionResponse',
	type: 'object',
	required: ['events', 'result'],
	properties: {
		events: {
			fieldNumber: 1,
			type: 'array',
			items: {
				...eventSchema,
			},
		},
		result: {
			fieldNumber: 2,
			dataType: 'sint32',
		},
	},
};

export const commitRequestSchema = {
	$id: '/abi/commitRequest',
	type: 'object',
	required: ['contextID', 'stateRoot', 'expectedStateRoot', 'dryRun'],
	properties: {
		contextID: {
			fieldNumber: 1,
			dataType: 'bytes',
		},
		stateRoot: {
			fieldNumber: 2,
			dataType: 'bytes',
		},
		expectedStateRoot: {
			fieldNumber: 3,
			dataType: 'bytes',
		},
		dryRun: {
			fieldNumber: 4,
			dataType: 'boolean',
		},
	},
};

export const commitResponseSchema = {
	$id: '/abi/commitResponse',
	type: 'object',
	properties: {
		stateRoot: {
			fieldNumber: 1,
			dataType: 'bytes',
		},
	},
};

export const revertRequestSchema = {
	$id: '/abi/revertRequest',
	type: 'object',
	required: ['contextID', 'stateRoot', 'expectedStateRoot'],
	properties: {
		contextID: {
			fieldNumber: 1,
			dataType: 'bytes',
		},
		stateRoot: {
			fieldNumber: 2,
			dataType: 'bytes',
		},
		expectedStateRoot: {
			fieldNumber: 3,
			dataType: 'bytes',
		},
	},
};

export const revertResponseSchema = {
	$id: '/abi/revertResponse',
	type: 'object',
	required: ['stateRoot'],
	properties: {
		stateRoot: {
			fieldNumber: 1,
			dataType: 'bytes',
		},
	},
};

export const finalizeRequestSchema = {
	$id: '/abi/finalizeRequest',
	type: 'object',
	required: ['finalizedHeight'],
	properties: {
		finalizedHeight: {
			fieldNumber: 1,
			dataType: 'uint32',
		},
	},
};

export const finalizeResponseSchema = {
	$id: '/abi/finalizeResponse',
	type: 'object',
	properties: {},
};

export const clearRequestSchema = {
	$id: '/abi/clearRequest',
	type: 'object',
	required: [],
	properties: {},
};

export const clearResponseSchema = {
	$id: '/abi/clearResponse',
	type: 'object',
	properties: {},
};

export const metadataRequestSchema = {
	$id: '/abi/metadataRequest',
	type: 'object',
	properties: {},
};

export const metadataResponseSchema = {
	$id: '/abi/metadataResponse',
	type: 'object',
	required: ['data'],
	properties: {
		data: {
			fieldNumber: 1,
			dataType: 'bytes',
		},
	},
};

export const queryRequestSchema = {
	$id: '/abi/queryRequest',
	type: 'object',
	required: ['method', 'params', 'header'],
	properties: {
		method: {
			fieldNumber: 1,
			dataType: 'string',
		},
		params: {
			fieldNumber: 2,
			dataType: 'bytes',
		},
		header: {
			fieldNumber: 3,
			...blockHeaderSchema,
		},
	},
};

export const queryResponseSchema = {
	$id: '/abi/queryResponse',
	type: 'object',
	required: ['data'],
	properties: {
		data: {
			fieldNumber: 1,
			dataType: 'bytes',
		},
	},
};

export const proveRequestSchema = {
	$id: '/abi/proveRequest',
	type: 'object',
	required: ['stateRoot', 'keys'],
	properties: {
		stateRoot: {
			fieldNumber: 1,
			dataType: 'bytes',
		},
		keys: {
			fieldNumber: 2,
			type: 'array',
			items: {
				dataType: 'bytes',
			},
		},
	},
};

export const proveResponseSchema = {
	$id: '/abi/proveResponse',
	type: 'object',
	required: ['proof'],
	properties: {
		proof: {
			fieldNumber: 1,
			type: 'object',
			required: ['siblingHashes', 'queries'],
			properties: {
				siblingHashes: {
					fieldNumber: 1,
					type: 'array',
					items: {
						dataType: 'bytes',
					},
				},
				queries: {
					fieldNumber: 2,
					type: 'array',
					items: {
						type: 'object',
						required: ['key', 'value', 'bitmap'],
						properties: {
							key: {
								fieldNumber: 1,
								dataType: 'bytes',
							},
							value: {
								fieldNumber: 2,
								dataType: 'bytes',
							},
							bitmap: {
								fieldNumber: 3,
								dataType: 'bytes',
							},
						},
					},
				},
			},
		},
	},
};

export const ipcRequestSchema = {
	$id: '/abi/ipcRequest',
	type: 'object',
	required: ['id', 'method', 'params'],
	properties: {
		id: {
			fieldNumber: 1,
			dataType: 'uint64',
		},
		method: {
			fieldNumber: 2,
			dataType: 'string',
		},
		params: {
			fieldNumber: 3,
			dataType: 'bytes',
		},
	},
};

export const ipcResponseSchema = {
	$id: '/abi/ipcResponse',
	type: 'object',
	required: ['id', 'success', 'error', 'result'],
	properties: {
		id: {
			fieldNumber: 1,
			dataType: 'uint64',
		},
		success: {
			fieldNumber: 2,
			dataType: 'boolean',
		},
		error: {
			type: 'object',
			fieldNumber: 3,
			required: ['message'],
			properties: {
				message: {
					fieldNumber: 1,
					dataType: 'string',
				},
			},
		},
		result: {
			fieldNumber: 4,
			dataType: 'bytes',
		},
	},
};
