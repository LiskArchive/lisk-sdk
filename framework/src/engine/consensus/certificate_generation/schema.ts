/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export const unsignedCertificateSchema = {
	$id: '/consensus/certificate',
	type: 'object',
	required: ['blockID', 'height', 'timestamp', 'stateRoot', 'validatorsHash'],
	properties: {
		blockID: {
			dataType: 'bytes',
			fieldNumber: 1,
		},
		height: {
			dataType: 'uint32',
			fieldNumber: 2,
		},
		timestamp: {
			dataType: 'uint32',
			fieldNumber: 3,
		},
		stateRoot: {
			dataType: 'bytes',
			fieldNumber: 4,
		},
		validatorsHash: {
			dataType: 'bytes',
			fieldNumber: 5,
		},
	},
};

export const certificateSchema = {
	$id: '/consensus/certificate',
	type: 'object',
	required: [
		'blockID',
		'height',
		'timestamp',
		'stateRoot',
		'validatorsHash',
		'aggregationBits',
		'signature',
	],
	properties: {
		blockID: {
			dataType: 'bytes',
			fieldNumber: 1,
		},
		height: {
			dataType: 'uint32',
			fieldNumber: 2,
		},
		timestamp: {
			dataType: 'uint32',
			fieldNumber: 3,
		},
		stateRoot: {
			dataType: 'bytes',
			fieldNumber: 4,
		},
		validatorsHash: {
			dataType: 'bytes',
			fieldNumber: 5,
		},
		aggregationBits: {
			dataType: 'bytes',
			fieldNumber: 6,
		},
		signature: {
			dataType: 'bytes',
			fieldNumber: 7,
		},
	},
};

export const singleCommitSchema = {
	$id: '/consensus/singleCommit',
	type: 'object',
	required: ['blockID', 'height', 'validatorAddress', 'certificateSignature'],
	properties: {
		blockID: {
			dataType: 'bytes',
			fieldNumber: 1,
			minLength: 32,
			maxLength: 32,
		},
		height: {
			dataType: 'uint32',
			fieldNumber: 2,
		},
		validatorAddress: {
			dataType: 'bytes',
			fieldNumber: 3,
			format: 'lisk32',
		},
		certificateSignature: {
			dataType: 'bytes',
			fieldNumber: 4,
			minLength: 96,
			maxLength: 96,
		},
	},
};

export const singleCommitsNetworkPacketSchema = {
	$id: '/consensus/singleCommitsNetworkPacket',
	type: 'object',
	required: ['commits'],
	properties: {
		commits: {
			type: 'array',
			fieldNumber: 1,
			items: {
				dataType: 'bytes',
			},
		},
	},
};

export const aggregateCommitSchema = {
	$id: '/consensus/aggregateCommit',
	type: 'object',
	required: ['height', 'aggregationBits', 'certificateSignature'],
	properties: {
		height: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
		aggregationBits: {
			dataType: 'bytes',
			fieldNumber: 2,
		},
		certificateSignature: {
			dataType: 'bytes',
			fieldNumber: 3,
		},
	},
};
