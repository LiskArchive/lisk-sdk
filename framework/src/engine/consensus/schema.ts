/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { SingleCommit } from './certificate_generation/types';
import { singleCommitSchema } from './certificate_generation/schema';

export interface RPCBlocksByIdData {
	readonly blockId: Buffer;
}

export const getBlocksFromIdRequestSchema = {
	$id: 'lisk/getBlocksFromIdRequest',
	title: 'Get Blocks From Id Request',
	type: 'object',
	required: ['blockId'],
	properties: {
		blockId: {
			fieldNumber: 1,
			dataType: 'bytes',
			minLength: 32,
			maxLength: 32,
		},
	},
};

export const getBlocksFromIdResponseSchema = {
	$id: 'lisk/getBlocksFromIdResponse',
	title: 'Get Blocks From Id Response',
	type: 'object',
	required: ['blocks'],
	properties: {
		blocks: {
			type: 'array',
			fieldNumber: 1,
			items: {
				dataType: 'bytes',
			},
		},
	},
};

export const getHighestCommonBlockRequestSchema = {
	$id: 'lisk/getHighestCommonBlockRequest',
	title: 'Get Highest Common Block Request',
	type: 'object',
	required: ['ids'],
	properties: {
		ids: {
			type: 'array',
			fieldNumber: 1,
			minItems: 1,
			items: {
				dataType: 'bytes',
				minLength: 32,
				maxLength: 32,
			},
		},
	},
};

export interface RPCHighestCommonBlockRequest {
	readonly ids: Buffer[];
}

export const getHighestCommonBlockResponseSchema = {
	$id: 'lisk/getHighestCommonBlockResponse',
	title: 'Get Highest Common Block Response',
	type: 'object',
	required: ['id'],
	properties: {
		id: {
			dataType: 'bytes',
			fieldNumber: 1,
			minLength: 32,
			maxLength: 32,
		},
	},
};

export interface RPCHighestCommonBlockResponse {
	readonly id: Buffer;
}

export interface EventPostBlockData {
	readonly block: Buffer;
}

export const postBlockEventSchema = {
	$id: 'lisk/postBlockEvent',
	title: 'Post Block Event',
	type: 'object',
	required: ['block'],
	properties: {
		block: {
			dataType: 'bytes',
			fieldNumber: 1,
		},
	},
};

export interface SingleCommitData {
	singleCommit: SingleCommit;
}

export const getSingleCommitEventSchema = {
	$id: 'lisk/getSingleCommitEvent',
	title: 'Get SingleCommit Event',
	type: 'object',
	required: ['singleCommit'],
	properties: {
		singleCommit: {
			type: 'object',
			fieldNumber: 1,
			required: ['blockID', 'height', 'validatorAddress', 'certificateSignature'],
			properties: {
				...singleCommitSchema.properties,
			},
		},
	},
};
