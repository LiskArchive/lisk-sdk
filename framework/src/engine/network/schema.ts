/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export const customNodeInfoSchema = {
	$id: '/nodeInfo/custom',
	type: 'object',
	properties: {
		height: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
		maxHeightPrevoted: {
			dataType: 'uint32',
			fieldNumber: 2,
		},
		blockVersion: {
			dataType: 'uint32',
			fieldNumber: 3,
		},
		lastBlockID: {
			dataType: 'bytes',
			fieldNumber: 4,
		},
	},
};
