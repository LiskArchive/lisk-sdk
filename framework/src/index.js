/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

'use strict';

const Application = require('./application/application');
const version = require('./version');
const defaultConfigurator = require('./application/default_configurator');
const systemDirs = require('./application/system_dirs');
const BaseModule = require('./modules/base_module');

module.exports = {
	Application,
	version,
	systemDirs,
	configurator: defaultConfigurator,
	BaseModule,
};
