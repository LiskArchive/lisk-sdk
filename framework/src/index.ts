/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export { Transaction } from '@liskhq/lisk-chain';
export { BaseModule, BaseAsset } from './modules';
export { Application } from './application';
export { version } from './version';
export { systemDirs } from './application/system_dirs';
export { BasePlugin, BlockHeaderJSON, PluginInfo, PluginCodec } from './plugins/base_plugin';
export { IPCChannel } from './controller/channels';
export type { BaseChannel } from './controller/channels';
export type { EventsArray, EventInfoObject } from './controller/event';
export type { ActionsDefinition } from './controller/action';
export * from './types';
