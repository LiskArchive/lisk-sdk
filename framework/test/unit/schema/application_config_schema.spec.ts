/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { validator } from '@liskhq/lisk-validator';
import { objects } from '@liskhq/lisk-utils';
import { applicationConfigSchema } from '../../../src/schema';

describe('schema/application_config_schema.js', () => {
	it('application config schema must match to the snapshot.', () => {
		expect(applicationConfigSchema).toMatchSnapshot();
	});

	it('should validate the defined schema', () => {
		expect(
			() => validator.validateSchema(applicationConfigSchema)
		).not.toThrow()
	});

	it('should validate if module properties are objects', () => {
		const config = objects.cloneDeep(applicationConfigSchema.default);
		config.genesis.modules = { myModule: { myProp: 1 } };

		expect(
			() => validator.validate(applicationConfigSchema, config)
		).not.toThrow()
	});

	it('should not validate if module properties are not objects', () => {
		const config = objects.cloneDeep(applicationConfigSchema.default);
		config.genesis.modules = { myModule: 10 };

		expect(
			() => validator.validate(applicationConfigSchema, config)
		).toThrow("Property '.genesis.modules.myModule' should be of type 'object'")
	});

	it('should not validate if module properties are not valid format', () => {
		const config = objects.cloneDeep(applicationConfigSchema.default);
		config.genesis.modules = { 'my-custom-module': { myProp: 1 } };

		expect(
			() => validator.validate(applicationConfigSchema, config)
		).toThrow("must match pattern \"^[a-zA-Z][a-zA-Z0-9_]*$\"\nproperty name must be valid")
	});
});
