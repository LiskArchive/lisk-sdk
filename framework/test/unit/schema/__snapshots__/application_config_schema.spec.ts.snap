// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema/application_config_schema.js application config schema must match to the snapshot. 1`] = `
{
  "$id": "#/config",
  "additionalProperties": false,
  "default": {
    "generator": {
      "keys": {},
    },
    "genesis": {
      "bftBatchSize": 103,
      "block": {
        "fromFile": "./config/genesis_block.blob",
      },
      "blockTime": 10,
      "maxTransactionsSize": 15360,
    },
    "legacy": {
      "brackets": [],
      "sync": false,
    },
    "modules": {},
    "network": {
      "port": 7667,
      "seedPeers": [],
      "version": "1.0",
    },
    "plugins": {},
    "rpc": {
      "host": "127.0.0.1",
      "modes": [
        "ipc",
      ],
      "port": 7887,
    },
    "system": {
      "dataPath": "~/.lisk/beta-sdk-app",
      "keepEventsForHeights": 300,
      "logLevel": "info",
      "version": "0.1.0",
    },
    "transactionPool": {
      "maxTransactions": 4096,
      "maxTransactionsPerAccount": 64,
      "minEntranceFeePriority": "0",
      "minReplacementFeeDifference": "10",
      "transactionExpiryTime": 10800000,
    },
  },
  "properties": {
    "generator": {
      "properties": {
        "keys": {
          "properties": {
            "fromFile": {
              "description": "Path to a file which stores keys",
              "type": "string",
            },
          },
          "type": "object",
        },
      },
      "required": [
        "keys",
      ],
      "type": "object",
    },
    "genesis": {
      "additionalProperties": false,
      "properties": {
        "bftBatchSize": {
          "description": "The length of a round",
          "minimum": 1,
          "type": "number",
        },
        "block": {
          "oneOf": [
            {
              "properties": {
                "blob": {
                  "format": "hex",
                  "type": "string",
                },
                "fromFile": {
                  "type": "string",
                },
              },
              "required": [
                "fromFile",
              ],
            },
            {
              "properties": {
                "blob": {
                  "format": "hex",
                  "type": "string",
                },
                "fromFile": {
                  "type": "string",
                },
              },
              "required": [
                "blob",
              ],
            },
          ],
          "type": "object",
        },
        "blockTime": {
          "description": "Slot time interval in seconds",
          "minimum": 2,
          "type": "number",
        },
        "chainID": {
          "description": "The unique name of the chain as a string encoded in Hex format",
          "format": "hex",
          "type": "string",
        },
        "maxTransactionsSize": {
          "description": "Maximum number of transactions allowed per block",
          "maximum": 30720,
          "minimum": 10240,
          "type": "integer",
        },
      },
      "required": [
        "block",
        "blockTime",
        "bftBatchSize",
        "chainID",
        "maxTransactionsSize",
      ],
      "type": "object",
    },
    "legacy": {
      "properties": {
        "brackets": {
          "items": {
            "properties": {
              "snapshotBlockID": {
                "format": "hex",
                "type": "string",
              },
              "snapshotHeight": {
                "minimum": 0,
                "type": "number",
              },
              "startHeight": {
                "minimum": 0,
                "type": "number",
              },
            },
            "required": [
              "startHeight",
              "snapshotHeight",
              "snapshotBlockID",
            ],
            "type": "object",
          },
          "type": "array",
        },
        "sync": {
          "type": "boolean",
        },
      },
      "required": [
        "brackets",
      ],
      "type": "object",
    },
    "modules": {
      "additionalProperties": {
        "type": "object",
      },
      "propertyNames": {
        "pattern": "^[a-zA-Z][a-zA-Z0-9_]*$",
      },
      "type": "object",
    },
    "network": {
      "properties": {
        "advertiseAddress": {
          "type": "boolean",
        },
        "blacklistedIPs": {
          "items": {
            "format": "ip",
            "type": "string",
          },
          "type": "array",
        },
        "fixedPeers": {
          "items": {
            "properties": {
              "ip": {
                "format": "ip",
                "type": "string",
              },
              "port": {
                "maximum": 65535,
                "minimum": 1,
                "type": "integer",
              },
            },
            "type": "object",
          },
          "maximum": 4,
          "type": "array",
        },
        "host": {
          "format": "ip",
          "type": "string",
        },
        "maxInboundConnections": {
          "type": "integer",
        },
        "maxOutboundConnections": {
          "type": "integer",
        },
        "port": {
          "maximum": 65535,
          "minimum": 1,
          "type": "integer",
        },
        "seedPeers": {
          "items": {
            "properties": {
              "ip": {
                "format": "ipOrFQDN",
                "type": "string",
              },
              "port": {
                "maximum": 65535,
                "minimum": 1,
                "type": "integer",
              },
            },
            "type": "object",
          },
          "type": "array",
        },
        "version": {
          "format": "networkVersion",
          "type": "string",
        },
        "whitelistedPeers": {
          "items": {
            "properties": {
              "ip": {
                "format": "ip",
                "type": "string",
              },
              "port": {
                "maximum": 65535,
                "minimum": 1,
                "type": "integer",
              },
            },
            "type": "object",
          },
          "type": "array",
        },
        "wsMaxPayload": {
          "maximum": 3048576,
          "type": "integer",
        },
      },
      "required": [
        "seedPeers",
      ],
      "type": "object",
    },
    "plugins": {
      "type": "object",
    },
    "rpc": {
      "properties": {
        "host": {
          "type": "string",
        },
        "modes": {
          "items": {
            "enum": [
              "ipc",
              "ws",
              "http",
            ],
            "type": "string",
          },
          "type": "array",
          "uniqueItems": true,
        },
        "port": {
          "maximum": 65535,
          "minimum": 1024,
          "type": "number",
        },
      },
      "required": [
        "modes",
        "host",
        "port",
      ],
      "type": "object",
    },
    "system": {
      "properties": {
        "dataPath": {
          "type": "string",
        },
        "keepEventsForHeights": {
          "type": "integer",
        },
        "logLevel": {
          "enum": [
            "trace",
            "debug",
            "info",
            "warn",
            "error",
            "fatal",
            "none",
          ],
          "type": "string",
        },
        "version": {
          "format": "version",
          "type": "string",
        },
      },
      "required": [
        "version",
        "dataPath",
        "logLevel",
        "keepEventsForHeights",
      ],
      "type": "object",
    },
    "transactionPool": {
      "properties": {
        "maxTransactions": {
          "minimum": 1,
          "type": "integer",
        },
        "maxTransactionsPerAccount": {
          "minimum": 1,
          "type": "integer",
        },
        "minEntranceFeePriority": {
          "format": "uint64",
          "type": "string",
        },
        "minReplacementFeeDifference": {
          "format": "uint64",
          "type": "string",
        },
        "transactionExpiryTime": {
          "minimum": 60000,
          "type": "integer",
        },
      },
      "type": "object",
    },
  },
  "required": [
    "system",
    "rpc",
    "network",
    "modules",
    "plugins",
    "genesis",
  ],
  "type": "object",
}
`;
